normandy/recipes/utils.py

```diff
@@ -46,3 +54,68 @@ def deterministic_sample(rate, inputs):
     assert len(input_hash) == 64
 
     return input_hash < sample_point
+
+
+class Autographer:
+    """
+    Interacts with an Autograph service.
+
+    If Autograph signing is not configured using `settings.AUTOGRAPH`,
+    raises `ImproperlyConfigured`. If the Autograph server can't be reached
+    or returns an HTTP error, an error will be thrown by `requests`.
+    """
+
+    def __init__(self):
+        self.check_config()
+        self._session = None
+
+    @property
+    def session(self):
+        if self._session is None:
+            self._session = requests.Session()
+            self._session.auth = HawkAuth(
+                id=str(settings.AUTOGRAPH_HAWK_ID),
+                key=str(settings.AUTOGRAPH_HAWK_SECRET_KEY))
+        return self._session
+
+    def check_config(self):
+        required_keys = ['URL', 'HAWK_ID', 'HAWK_SECRET_KEY']
+        for key in required_keys:
+            if getattr(settings, 'AUTOGRAPH_' + key) is None:
+                msg = 'set settings.AUTOGRAPH_{} to use action signatures'.format(key)
+                raise ImproperlyConfigured(msg)
+
+    def sign_data(self, content_list):
+        """
+        Fetches Signatures objects from Autograph for each item in `content_list`.
+
+        The items in `content_list` must be bytes objects.
+        """
+        ts = timezone.now()
+        url = '{}sign/data'.format(settings.AUTOGRAPH_URL)
```

Comment:
We've talked about this before. `urljoin` makes the situation worse, not better.

```python
>>> from urllib.parse import urljoin
>>> urljoin('https://example.com/api/1/', 'sign/data')
'https://example.com/api/1/sign/data' # good...
>>> urljoin('https://example.com/api/1/', '/sign/data')
'https://example.com/sign/data' # bad, only kept hostname
>>> urljoin('https://example.com/api/1', 'sign/data')
'https://example.com/api/sign/data' # bad, stripped off "1"
>>> urljoin('https://example.com/api/1', '/sign/data')
'https://example.com/sign/data' # bad, only kept hostname
```

We should probably write a helper util that does this.